tracing-subscriber/src/filter/directive.rs

```diff
@@ -0,0 +1,564 @@
+
+    pub(super) fn to_static(&self) -> Option<StaticDirective> {
+        if !self.is_static() {
+            return None;
+        }
+
+        // TODO(eliza): these strings are all immutable; we should consider
+        // `Arc`ing them to make this more efficient...
+        let field_names = self.fields.iter().map(field::Match::name).collect();
+
+        Some(StaticDirective {
+            target: self.target.clone(),
+            field_names,
+            level: self.level.clone(),
+        })
+    }
+
+    fn is_static(&self) -> bool {
+        !self.has_name() && !self.fields.iter().any(field::Match::has_value)
+    }
+
+    pub(super) fn is_dynamic(&self) -> bool {
+        self.has_name() || self.has_fields()
+    }
+
+    pub fn field_matcher(&self, meta: &Metadata) -> Option<field::CallsiteMatch> {
+        let fieldset = meta.fields();
+        let fields = self
+            .fields
+            .iter()
+            .filter_map(
+                |field::Match {
+                     ref name,
+                     ref value,
+                 }| {
+                    if let Some(field) = fieldset.field(name) {
+                        let value = value.as_ref().cloned()?;
+                        Some(Ok((field, value)))
+                    } else {
+                        Some(Err(()))
+                    }
+                },
+            )
+            .collect::<Result<FieldMap<_>, ()>>()
+            .ok()?;
+        Some(field::CallsiteMatch {
+            fields,
+            level: self.level.clone(),
+        })
+    }
+
+    pub(super) fn make_tables(
+        directives: impl IntoIterator<Item = Directive>,
+    ) -> (Dynamics, Statics) {
+        // TODO(eliza): this could be made more efficient...
+        let (dyns, stats): (BTreeSet<Directive>, BTreeSet<Directive>) =
+            directives.into_iter().partition(Directive::is_dynamic);
+        let statics = stats
+            .into_iter()
+            .filter_map(|d| d.to_static())
+            .chain(dyns.iter().filter_map(Directive::to_static))
+            .collect();
+        (Dynamics::from_iter(dyns), statics)
+    }
+}
+
+impl Match for Directive {
+    fn cares_about(&self, meta: &Metadata) -> bool {
+        // Does this directive have a target filter, and does it match the
+        // metadata's target?
+        if let Some(ref target) = self.target.as_ref() {
+            if !meta.target().starts_with(&target[..]) {
+                return false;
+            }
+        }
+
+        // Do we have a name filter, and does it match the metadata's name?
+        // TODO(eliza): put name globbing here?
+        if let Some(ref name) = self.in_span {
+            if name != meta.name() {
+                return false;
+            }
+        }
+
+        // Does the metadata define all the fields that this directive cares about?
+        let fields = meta.fields();
+        for field in &self.fields {
+            if !fields.field(&field.name).is_some() {
```

Comment:
Is there a reason to prefer `!is_some()` to `is_none()`?

Reply:
nope, thanks for spotting that!